onnx/defs/shape_inference.h

```diff
@@ -291,85 +309,96 @@ inline void updateOutputShape(
-// to be a list of integers specifying a valid shape.
+// Infer shape of an output from the value of a specified attribute, which is
+// expected to be a list of integers specifying a valid shape.
 inline void propagateShapeFromAttributeToOutput(
     InferenceContext& ctx,
     const std::string& attributeName,
     size_t outputIndex) {
-
   auto attr_proto = ctx.getAttribute(attributeName);
-  if ((nullptr == attr_proto) ||
-      (!attr_proto->has_type()) ||
+  if ((nullptr == attr_proto) || (!attr_proto->has_type()) ||
       (attr_proto->type() != AttributeProto_AttributeType_INTS)) {
-	fail_shape_inference("Attribute ", attributeName, " should specify a shape");
+    fail_shape_inference(
+        "Attribute ", attributeName, " should specify a shape");
   }
   auto& int_list = attr_proto->ints();
   TensorShapeProto shape;
   for (auto dim_size : int_list) {
     if (dim_size < 0) {
-		fail_shape_inference("Negative values are not allowed in a shape specification");
-	}
+      fail_shape_inference(
+          "Negative values are not allowed in a shape specification");
+    }
     shape.add_dim()->set_dim_value(dim_size);
   }
 
   updateOutputShape(ctx, outputIndex, shape);
 }
 
-inline void bidirectionalBroadcastShapeInference(
-    const TensorShapeProto& shapeL,
-    const TensorShapeProto& shapeR,
+inline void multidirectionalBroadcastShapeInference(
+    const std::vector<const TensorShapeProto*>& shapes,
     TensorShapeProto& resultShape) {
-  int il = 0, ir = 0;
-  // copy prefix of shapeL, if shapeL is higher-rank
-  for (; il < shapeL.dim_size() - shapeR.dim_size(); ++il) {
-    *resultShape.add_dim() = shapeL.dim(il);
-  }
-  // copy prefix of shapeR, if shapeR is higher-rank
-  for (; ir < shapeR.dim_size() - shapeL.dim_size(); ++ir) {
-    *resultShape.add_dim() = shapeR.dim(ir);
+  int result_shape_size = 0;
+  // Get the result shape size.
+  for (size_t i = 0; i < shapes.size(); ++i) {
+    if (shapes[i]->dim_size() > result_shape_size) {
+      result_shape_size = shapes[i]->dim_size();
+    }
   }
 
-  for (; il < shapeL.dim_size(); ++il, ++ir) {
-    auto newdim = resultShape.add_dim();
-    if (shapeL.dim(il).has_dim_value() && shapeR.dim(ir).has_dim_value()) {
-      auto dimL = shapeL.dim(il).dim_value();
-      auto dimR = shapeR.dim(ir).dim_value();
-      if (dimL == dimR) {
-        newdim->set_dim_value(dimL);
-      } else if (dimL == 1) {
-        newdim->set_dim_value(dimR);
-      } else if (dimR == 1) {
-        newdim->set_dim_value(dimL);
-      } else {
-        fail_shape_inference("Incompatible dimensions");;
-      }
-    } else if (shapeL.dim(il).has_dim_value()) {
-      auto dimL = shapeL.dim(il).dim_value();
-      if (dimL == 1) {
-        *newdim = shapeR.dim(ir);
-      } else {
-        newdim->set_dim_value(dimL);
+  for (int i = 0; i < result_shape_size; ++i) {
+    int64_t dim_value = 1;
+    TensorShapeProto_Dimension symbolic_dim;
```

Comment:
What's the purpose of symbolic_dim? The logic seems incomplete.

Reply:
offline explained, symbolic dim is used,
1, if there's one symbolic dim and all others are 1, then symbolic dim will be used in result shape.
2. if there're symbolic dims and one dim value > 1, then the result shape will take the >1 value.
3. if there're more than 1 symbolic dim and all others are 1, then the result shape will not be set.